### ISSUE TITLE
Secret is lost when svc process crashes and restarted

### ISSUE THREAD

**Issue by gomoripeti** [rabbitmq/credentials-obfuscation#18]
When `credentials_obfuscation_svc` crashes and restarted with a fresh state it will have the initial pending-secret. If no caller calls `set_secret` again (and how would the caller knew it should) it will stay in this state going forward. It can result in:
- from this point on no data encrypted before the crash can be decrypted
- exposing this bug https://github.com/rabbitmq/credentials-obfuscation/pull/17

What if the secret (and probably other state as well) would be stored in persistent_term. encrypt/decrypt could be executed in the calling process, the gen_server is only kept for serialising state updates?

This would also prevent the secrets being logged at gen_server crash. Would this hurt observability?

Any feedback is appreciated.

**Comment by michaelklishin** [comment]
Supporting anonymous functions or MFAs to set the secret should be fine as long as we make sure their arguments themselves are never logged.

**Comment by gomoripeti** [comment]
what I meant is just wrapping a term into an anon fun (instead of encryption) as a way of obfuscation. Then it does not need encryption config and state/secret and no need for a gen_server process. Eg:

```
encrypt(Term) -> {encrypted, fun() -> Term end}.
decrypt({encrypted, Fun}) -> Fun().
```

Or is there any use-case/customer that needs actually encryption?
(One disadvantage I can think of is if the secret at a past point in time is known from some source eg it is explicitly configured, then encrypted secrets in old logs can still be decrypted for debugging purposes. There is no way to decrypt `#Fun<...>` just from the logs, without the actual term in the live VM being available)

**Comment by michaelklishin** [comment]
This option was considered but we decided to go with encryption. You cannot know how a future Erlang logger would potentially log anonymous functions differently.

**Comment by gomoripeti** [comment]
makes sense, thanks for the historic insight.

then in order to address the original issue, that if the svc process crashes, it loses the secret, would storing the state in persistent_term be acceptable? (now that only new enough versions of OTP are supported which all have persistent_term) (if yes, Im willing to prepare a PR)

**Comment by gomoripeti** [comment]
thanks Michal, I see this is complicated, so I refrain from major changes of this lib for now

Just to clarify the expectations against this lib
- I see in the README:
>this application cannot protect against heap dumping attacks and only helps avoid sensitive data appearing in log files.
- Also currently there is `credentials_obfuscation:secret()` to fetch the secret (which is the same as reading a well-known key from persistent_term).
 
Because of this existing API and the Readme I thought it is not a goal to give protection if someone has shell access to the node.

**Comment by michaelklishin** [comment]
Fair enough, we can use a persistent term, even though it would be susceptible to another type of attack: bogus plugins that would be able to access the term.

**Comment by michaelklishin** [comment]
@gomoripeti I asked half a dozen of engineers about this and we have a consensus that using a persistent term with a well-known key is fine. We just have to be careful to also pass the  fallback secret on the same way.
